src/fallback.ts

```diff
@@ -313,30 +325,90 @@ export class GrpcClient {
+        let url: string;
+        let data: string;
+        let httpMethod: string;
+
+        if (this.fallback === 'json') {
+          // REGAPIC: JSON over HTTP/1 with gRPC trancoding
+          headers['Content-Type'] = 'application/json';
+          const decodedRequest = method.resolvedRequestType.decode(requestData);
+          const requestJSON = method.resolvedRequestType.toObject(
+            // TODO: use toJSON instead of toObject
+            decodedRequest
+          );
+          const transcoded = transcode(requestJSON, method.parsedOptions);
+          if (!transcoded) {
+            throw new Error(
+              `Cannot build HTTP request for ${JSON.stringify(
+                requestJSON
+              )}, method: ${method.name}`
+            );
+          }
+          httpMethod = transcoded.httpMethod;
+          data = JSON.stringify(transcoded.data);
+          url = `${grpcFallbackProtocol}://${servicePath}:${servicePort}/${transcoded.url.replace(
+            /^\//,
+            ''
+          )}?${transcoded.queryString}`;
+        } else {
+          // gRPC-fallback: proto over HTTP/1
+          headers['Content-Type'] = 'application/x-protobuf';
+          httpMethod = 'post';
+          data = requestData;
+          url = `${grpcFallbackProtocol}://${servicePath}:${servicePort}/$rpc/${protoServiceName}/${rpcName}`;
+        }
 
         const fetch = isBrowser()
           ? // eslint-disable-next-line no-undef
             window.fetch
           : ((nodeFetch as unknown) as NodeFetchType);
-        fetch(url, {
+        const fetchRequest = {
           headers,
-          method: 'post',
-          body: requestData,
+          body: data,
+          method: httpMethod,
           signal: cancelSignal,
-        })
+        };
+        if (
+          httpMethod === 'get' ||
+          httpMethod === 'delete' ||
+          httpMethod === 'head'
+        ) {
+          delete fetchRequest['body'];
+        }
+        fetch(url, fetchRequest)
           .then((response: Response | nodeFetch.Response) => {
             return Promise.all([
               Promise.resolve(response.ok),
               response.arrayBuffer(),
             ]);
           })
           .then(([ok, buffer]: [boolean, Buffer | ArrayBuffer]) => {
-            if (!ok) {
-              const error = statusDecoder.decodeErrorFromBuffer(buffer);
-              throw error;
+            if (this.fallback === 'json') {
+              // REGAPIC: JSON over HTTP/1
+              // eslint-disable-next-line node/no-unsupported-features/node-builtins
+              const decodedString = new TextDecoder().decode(buffer);
+              const response = JSON.parse(decodedString);
+              if (!ok) {
+                const error = Object.assign(
+                  new Error(response['error']['message']),
+                  response.error
```

Comment:
same nit, might be worth factoring some of this logic into a helper.

Reply:
Left TODO, for the same reason.